src/main/java/sirius/biz/analytics/charts/explorer/EventTimeseries.java

```diff
@@ -0,0 +1,162 @@
+
+    /**
+     * Creates a new timeseries for the given event type.
+     *
+     * @param eventType the type of events to query
+     */
+    public EventTimeseries(Class<? extends Event> eventType) {
+        this.eventType = eventType;
+    }
+
+    /**
+     * Specifies which fields to query (to put in the SELECT clause) and how to process them.
+     *
+     * @param fieldsToSelect      one or more field to query (this will be put behind the SELECT in the generated query).
+     * @param rowProcessorFactory creates a row-processor, which is then supplied with all rows of the resulting query.
+     *                            This can create (and then fill) one or more {@link TimeseriesData} instances based
+     *                            on the query results.
+     * @return the timeseries computer itself for fluent method calls
+     */
+    public EventTimeseries<O> withSelect(String fieldsToSelect,
+                                         Function<Timeseries, BiConsumer<LocalDate, Function<String, Double>>> rowProcessorFactory) {
+        this.fieldsToSelect = fieldsToSelect;
+        this.rowProcessorFactory = rowProcessorFactory;
+
+        return this;
+    }
+
+    /**
+     * Provides a safe way of adding a parameterized WHERE clause.
+     *
+     * @param condition        the SQL to add. This must be a constant string. Variables can be referenced like <tt>${var}</tt>
+     *                         which are then provided by the <tt>filterCustomizer</tt>
+     * @param filterCustomizer uses {@link SQLQuery#set(String, Object)} to provide the variables used by
+     *                         <tt>condition</tt>
+     * @return the timeseries computer itself for fluent method calls
+     */
+    public EventTimeseries<O> withWhere(String condition, BiConsumer<O, SQLQuery> filterCustomizer) {
+        this.condition = condition;
+        this.filterCustomizer = filterCustomizer;
+
+        return this;
+    }
+
+    /**
+     * Permits to add a constant and parameter-free WHERE clause.
+     *
+     * @param condition the condition to add to the SQL
+     * @return the timeseries computer itself for fluent method calls
+     */
+    public EventTimeseries<O> withWhere(String condition) {
+        this.condition = condition;
+
+        return this;
+    }
+
+    /**
+     * Permits to add one or more additional GROUP BY clauses.
+     *
+     * @param groupBy the clauses to add
+     * @return the timeseries computer itself for fluent method calls
+     */
+    public EventTimeseries<O> withGroupBy(String groupBy) {
+        this.groupBy = groupBy;
+
+        return this;
+    }
+
+    @Override
+    public void compute(O object, Timeseries timeseries) throws Exception {
+        if (fieldsToSelect == null) {
+            throw new IllegalStateException("Missing call to withSelect");
+        }
+        String query = "SELECT " + fieldsToSelect + ", " + "MONTH(eventDate) as month, " + "YEAR(eventDate) as year";
```

Comment:
Isn't it potentially dangerous to build such queries by hand?

Reply:
It sort of is, hence the warning in the JavaDoc of the class. The other option would be to accept a complete SQL statement, however the group by and the where part would be the same for all queries...

Reply:
Already missed that sentence in the JavaDoc during review 😅
So I'm skeptical that future Implementers of that class remember this important security notice. But I don't have a better suggestion right away...

Reply:
Idea: make fieldsToSelect a list, which is filled with calls to addSelectField().
This method in turn parses field names passed with a pattern, and so avoiding injections.
This way the SQL generation would be somewhat safer.

Reply:
these field names are not names but things like `countIf(foo > bar) AS foobar)` which are very hard to parse and verify as correct. The approach is to eliminate the helpers and always provide and execute a whole query - an example will be in the follow-up PR.